cpp/include/cudf/join_new.hpp

```diff
@@ -0,0 +1,345 @@
+                                   std::vector<cudf::size_type> const& base_left_on,
+                                   std::vector<cudf::size_type> const& base_right_on,
+                                   Filter filter,
+                                   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+  template <typename Filter>
+  std::unique_ptr<table> full_join(cudf::table_view const& left,
+                                   cudf::table_view const& right,
+                                   std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+                                   std::vector<cudf::size_type> const& base_left_on,
+                                   std::vector<cudf::size_type> const& base_right_on,
+                                   Filter filter,
+                                   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+} //namespace sort_merge
+
+namespace hash {
+  template <typename Filter>
+  std::unique_ptr<table> inner_join(cudf::table_view const& left,
+                                    cudf::table_view const& right,
+                                    std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+                                    std::vector<cudf::size_type> const& base_left_on,
+                                    std::vector<cudf::size_type> const& base_right_on,
+                                    Filter filter,
+                                    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+  template <typename Filter>
+  std::unique_ptr<table> left_join(cudf::table_view const& left,
+                                   cudf::table_view const& right,
+                                   std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+                                   std::vector<cudf::size_type> const& base_left_on,
+                                   std::vector<cudf::size_type> const& base_right_on,
+                                   Filter filter,
+                                   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+  template <typename Filter>
+  std::unique_ptr<table> full_join(cudf::table_view const& left,
+                                   cudf::table_view const& right,
+                                   std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+                                   std::vector<cudf::size_type> const& base_left_on,
+                                   std::vector<cudf::size_type> const& base_right_on,
+                                   Filter filter,
+                                   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+} //namespace hash
+
+//
+//  Option 3:  Something completely different
```

Comment:
Internally it's probably going to look more like this anyway so that we can reuse code more efficiently across kernels and host calls.

Reply:
Abandoned this approach.  If a caller needs something like this they can reach into the detail namespace.